Mironov_K - Task03

kish-mish opened the pull request.

kish-mish committed:
task03.1

Print statistic about commentators YT

kish-mish committed:
task03.2

del file chat

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+import os
+
+
+def pars_user_line(str_line: str) -> (str, str):
```
pars -> parse

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def pars_user_line(str_line: str) -> (str, str):
+    index = [i for i, char in enumerate(str_line) if char == ";"]
```
Это можно сделать намного проще с помощью `split(";")`

kish-mish replied:
Просто я посчитал, что лучше создать дополнительный список с индексами ";", чем дополнительный список со строками, где большинство элементов пустые

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def gen_user_dict(path_file: str) -> dict:
+    list_lines = list(filter(lambda it: it[0].isdigit(), open(path_file, "rt").readlines()))
```
Зачем здесь к списку преобразование?

kish-mish replied:
По сути, ненужное преобразование, забыл убрать после отладки

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def gen_user_dict(path_file: str) -> dict:
+    list_lines = list(filter(lambda it: it[0].isdigit(), open(path_file, "rt").readlines()))
```
open(path_file, "rt").readlines() - лучше вынести отдельно

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+def gen_user_dict(path_file: str) -> dict:
+    list_lines = list(filter(lambda it: it[0].isdigit(), open(path_file, "rt").readlines()))
+    dict_result = dict(map(pars_user_line, list_lines))
```
на мой взгляд, через list-comprehension было бы читабельней + можно там же фильтр поставить на пустую строку

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+    list_message = [it.split("\n?") for it in open(path_file, "rt").read().split("\n\n")]
+    dict_result = dict()
+    for it in list_message:
+        dict_result.setdefault(it[0], list())
```
setdefault возвращает добавленный элемент или полученный элемент

xthebat replied:
it[0] лучше вытаскивать в переменную, чтобы было по имени понятно, что это такое

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def gen_chat_dict(path_file: str) -> dict:
+    list_message = [it.split("\n?") for it in open(path_file, "rt").read().split("\n\n")]
```
это лучше разделить на несколько строк

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def gen_chat_dict(path_file: str) -> dict:
+    list_message = [it.split("\n?") for it in open(path_file, "rt").read().split("\n\n")]
```
что такое \n? - откуда '?'

kish-mish replied:
При чтении файла в ANSI формате выводится вопросительный знак перед комментарием

kish-mish replied:
![image](https://user-images.githubusercontent.com/55107877/158019911-58d4748d-fee2-4a91-b961-191d1f00253b.png)

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+    list_message = [it.split("\n?") for it in open(path_file, "rt").read().split("\n\n")]
+    dict_result = dict()
+    for it in list_message:
+        dict_result.setdefault(it[0], list())
```
за setdefault 🥇

kish-mish replied:
Спасибо, только так и не понял, правильно ли я использовал это. Перед этим вы оставили ревью с объяснением вывода функции.

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+def print_chat(user_dict: dict, chat_dict: dict):
+    for item in chat_dict:
+        print(item + " aka " + user_dict.get(item, "not declared"))
+        print("    Кол-во сообщений: " + str(len(chat_dict[item])))
```
используйте интерполяцию строк

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+
+def print_chat(user_dict: dict, chat_dict: dict):
+    for item in chat_dict:
+        print(item + " aka " + user_dict.get(item, "not declared"))
```
используйте интерполяцию строк, читать формирование строки через склейку намного сложнее из-за огромного количеств +

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -0,0 +1,65 @@
+    if not os.path.exists(args[2]):
+        sys.exit("Chat file is not exist")
+
+    if args[1][-4:] != ".csv":
```
так не надо проверять расширение файла, для этого есть функция:
https://stackoverflow.com/questions/541390/extracting-extension-from-filename-in-python

xthebat reviewed (commented):
Score: 
- ДЗ3: 2/4
- ДЗ4: 0/4

kish-mish committed:
task04.1

Выполнил сначала третье дз через comprehension, но пришлось перелопатить под исключения

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
+        return {}
+    else:
+        list_users = (it.split(";") for it in list_lines)
+        dict_result = {it[3]: it[1] for it in list_users if it[0].isdigit() and it[3] != ""}
```
3 и 1 - лучше вынести в константы, REAL_NAME_COLUMN, YT_NAME_COLUMN

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
 
 
-def main(args: list):
+def gen_user_dict(path_file: str) -> dict:
```
Опять (возможно тебе не писал по csv это замечание) смешались в кучу люди и кони, все подряд.

Если функция парсит csv, то ее результат должен быть какое-то представление таблицы, без привязки к логике работы программы. То есть на вход файл, на выход в самом простом случае List[List[str]], либо что-то что может представлять двумерную таблицу.

Дальше уже пишем отдельную функцию, которая построит словарь меппинга имен пользователя из этой таблицы.

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
+    except FileExtError as error:
+        print(f"File extention error: {error.file} is not {error.ext}")
+        return {}
+    else:
```
За else 👍

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
+        list_lines = open(path_file, "rt").readlines()
+    except FileNotFoundError as error:
+        print(error)
+        return {}
```
Согласен, что в комментариях обсуждали, что можно продолжать работу, даже если файл csv-битый, но:

![Uploading image.png…]()

kish-mish replied:
Картинка не отобржается

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
+        list_chat = open(path_file, mode="rt", encoding="UTF-16").read().split("\n\n")
+        for it in list_chat:
+            if it.count("\n\u200b") != 1:
+                raise ChatParseError(it)
```
Кидать ошибку, чтобы её поймать в той же функции ... ну такое, тут же можно сделать sys.exit(...) по логике остальной программы.

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
+        list_message = [it.split("\n\u200b") for it in list_chat]
+        # При копировании чата с ютуба перед комметарием есть дополнительный символ юникода: пробел нулевой ширины
+    except FileNotFoundError as error:
+        sys.exit(error)
```
Реализация отлова ошибок таким паттерном должна быть на самом верхнем уровне.

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -2,64 +2,90 @@
+        sys.exit(f"Error format file {path_file} in message:\n {error.line}")
+    else:
+        dict_result = dict()
+        for it in list_message:
```
🥇

xthebat reviewed (commented):
Score: 3/4 (разделите парсер csv и создание меппинга на две функции)
Score: 1/4

kish-mish committed:
task04.2

Divide gen_user_dict in parse_csv_file and gen_user_dict

xthebat commented on Mironov_K/task03/main.py:
```diff
@@ -18,50 +18,48 @@ def __init__(self, line: str):
+def parse_csv_file(path_file: str) -> list:
+    if os.path.splitext(path_file)[1] != ".csv":
+        raise FileExtError(path_file, "csv")
+    list_lines = open(path_file, "rt").readlines()
```
Файл скорее всего останется открытым, нужно либо явно закрывать с помощью `.close()`, либо использовать `with`

xthebat reviewed (approved):
ДЗ3: 4/4
ДЗ4: 4/4
